File: app/src/main/java/com/rowland/delivery/features/dash/presentation/viewmodels/category/CategoryViewModel.java
package com.rowland.delivery.features.dash.presentation.viewmodels.category;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.util.Log;

import com.rowland.delivery.features.dash.domain.models.category.Category;
import com.rowland.delivery.features.dash.domain.usecases.category.CreateCategoryUseCase;
import com.rowland.delivery.features.dash.domain.usecases.category.LoadCategoriesUseCase;
import com.rowland.delivery.features.dash.presentation.viewmodels.SharedViewModel;

import java.util.List;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Rowland on 5/14/2018.
 */

public class CategoryViewModel extends SharedViewModel<Category> {

    private final MutableLiveData<String> firebaseUserUid = new MutableLiveData<>();

    private final LoadCategoriesUseCase loadCategoriesUseCase;
    private final CreateCategoryUseCase createCategoryUseCase;

    public CategoryViewModel(LoadCategoriesUseCase loadCategoriesUseCase, CreateCategoryUseCase createCategoryUseCase) {
        this.loadCategoriesUseCase = loadCategoriesUseCase;
        this.createCategoryUseCase = createCategoryUseCase;
    }

    public LiveData<List<Category>> getDataList() {
        loadCategoriesUseCase.loadCategories(firebaseUserUid.getValue())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(categories -> {
                    dataList.setValue(categories);
                },throwable -> {
                    Log.d(com.rowland.delivery.features.dash.presentation.viewmodels.category.CategoryViewModel.class.getSimpleName(), "Error Retrieving List: "+throwable);
                },() -> {
                    Log.d(com.rowland.delivery.features.dash.presentation.viewmodels.category.CategoryViewModel.class.getSimpleName(), "Done Retrieving");
                });
        return dataList;
    }

    public void setFirebaseUserUid(String userUid) {
        firebaseUserUid.setValue(userUid);
    }

    public Single<Category> saveCategory(Category category) {
        return createCategoryUseCase.createCategory(category, firebaseUserUid.getValue());
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        // ToDo: clean subscriptions
    }
}

File: app/src/main/java/com/rowland/delivery/features/dash/presentation/viewmodels/category/CategoryViewModel.kt
package com.rowland.delivery.features.dash.presentation.viewmodels.category

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.rowland.delivery.features.dash.domain.models.category.CategoryEntity
import com.rowland.delivery.features.dash.domain.usecases.category.CreateCategoryUseCase
import com.rowland.delivery.features.dash.domain.usecases.category.LoadCategoriesUseCase
import com.rowland.delivery.features.dash.presentation.viewmodels.SharedViewModel
import com.rowland.delivery.features.dash.presentation.viewmodels.ViewEvent
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by Rowland on 5/14/2018.
 */

class CategoryViewModel @Inject constructor(private val loadCategoriesUseCase: LoadCategoriesUseCase, private val createCategoryUseCase: CreateCategoryUseCase) : SharedViewModel<CategoryEntity>() {

    init {
        loadCategoriesUseCase.loadCategories(FirebaseAuth.getInstance().currentUser!!.uid)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ categories ->
                    dataList.value = categories
                    updateState { it.copy(event = ViewEvent.Success()) }
                }, { throwable ->
                    updateState { it.copy(event = ViewEvent.Error()) }
                }) {
                    Log.d(CategoryViewModel::class.java.simpleName, "Done Retrieving")
                }
    }


    fun saveCategory(categoryEntity: CategoryEntity): Single<CategoryEntity> {
        return createCategoryUseCase.createCategory(categoryEntity, FirebaseAuth.getInstance().currentUser!!.uid)
    }
}
